File: src/binary-search/BOJ1920.ts
export {};
// https://www.acmicpc.net/problem/1920
// 수찾기
const inputStrings = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "./input.txt")
  .toString()
  .trim()
  .split("\n");

const testCases = inputStrings[3].split(" ").map(Number);
const numbers = inputStrings[1]
  .split(" ")
  .map(Number)
  .sort((a, b) => a - b);

let answer = "";
for (let i = 0; i < testCases.length; i++) {
  answer += `${solution(testCases[i])}\n`;
}
console.log(answer);

function solution(target: number) {
  let start = 0;
  let end = numbers.length - 1;
  while (start <= end) {
    const middle = Math.round((start + end) / 2);
    if (numbers[middle] < target) start = middle + 1;
    else if (numbers[middle] === target) return 1;
    else end = middle - 1;
  }

  return 0;
}

File: src/binary-search/BOJ1920.py
# 파이썬 sort 시간복잡도 (O NlogN)
# list.sort()
# list.sort(reverse=True)
# 기본값은 오름차순, reverse=True는 내림차순

# 선형탐색 O(n^2)
# 정렬 + 이분탐색: O(NlogN + MlogN)

# https://www.acmicpc.net/problem/1920
# 수 찾기

import sys
input = sys.stdin.readline


def binarySearch(start, end, target):
    while start <= end:
        mid = (start+end) // 2
        if numbers[mid] == target:
            return "1"
        elif numbers[mid] < target:
            start = mid + 1
        else:
            end = mid - 1
    return "0"


def solution():
    global numbers
    input()
    numbers = list(map(int, input().split()))
    numbers.sort()
    input()
    targets = list(map(int, input().split()))
    answer = []
    for target in targets:
        answer.append(binarySearch(0, len(numbers)-1, target))

    print("\n".join(answer))


solution()
